src/Raven.Server/Monitoring/ServerLimitsMonitor.cs

```diff
@@ -0,0 +1,214 @@
+            {
+                return;
+            }
+
+            _timer = new Timer(Run, null, CheckFrequency, CheckFrequency);
+        }
+
+        internal void Run(object state)
+        {
+            if (_notificationCenter.IsInitialized == false)
+                return;
+
+            if (Monitor.TryEnter(_runLock) == false)
+                return;
+
+            try
+            {
+                CheckMaxMapCountLimits();
+                CheckThreadsMaxLimits();
+                CheckPidMaxLimits();
+
+                if (_alerts.Count > 0)
+                {
+                    var warningMsg = string.Join(Environment.NewLine, _alerts);
+                    if (_logger.IsOperationsEnabled)
+                    {
+                        _logger.Operations($"Running close to OS limits detected:{Environment.NewLine}" + warningMsg);
+                    }
+
+                    var alert = AlertRaised.Create(
+                        null,
+                        "Running close to OS limits",
+                        warningMsg,
+                        AlertType.LowServerLimits,
+                        NotificationSeverity.Warning);
+
+                    _notificationCenter.Add(alert);
+                }
+            }
+            catch (Exception e)
+            {
+                if (_logger.IsOperationsEnabled)
+                    _logger.Operations($"Failed to run {nameof(ServerLimitsMonitor)}", e);
+            }
+            finally
+            {
+                _alerts.Clear();
+                Monitor.Exit(_runLock);
+            }
+        }
+
+        private void CheckPidMaxLimits()
+        {
+            var maxPidMax = GetMaxValueForLimit(pid_max);
+
+            using var proc = new Process();
+            proc.StartInfo.RedirectStandardOutput = true;
+            proc.StartInfo.RedirectStandardError = true;
+            proc.StartInfo.FileName = "ps";
+            proc.StartInfo.Arguments = @"-efL";
+            proc.Start();
+
+            proc.WaitForExit();
+
+            if (proc.ExitCode != 0)
+            {
+                throw new InvalidOperationException($"Executing '{proc.StartInfo.FileName} {proc.StartInfo.Arguments}' failed (exit code {proc.ExitCode}){Environment.NewLine}StandardOutput:{Environment.NewLine}{proc.StandardOutput.ReadToEnd()}{Environment.NewLine}StandardError:{Environment.NewLine}{proc.StandardError.ReadToEnd()}");
+            }
+
+            var currentPidMax = GetEolCount(proc.StandardOutput.BaseStream);
+
+            AddAlertIfNeeded(currentPidMax, maxPidMax, MaxPidMaxThreshold, pid_max, nameof(pid_max));
+        }
+
+        private void CheckThreadsMaxLimits()
+        {
+            var maxThreadsMax = GetMaxValueForLimit(threads_max);
+
+            var currentThreadsMax = 0L;
+            foreach (var _ in Directory.EnumerateDirectories("/proc/self/task", "*", SearchOption.TopDirectoryOnly))
```

Comment:
This is really expensive to do.
You do a lot of I/O, create a lot of strings, etc.

Reply:
See `/proc/self/stat` which has `num_threads`
https://man7.org/linux/man-pages/man5/proc.5.html

Reply:
done